### ISSUE TITLE
Cilium Endpoint generation Context deadline

### ISSUE THREAD

**Issue by eloycoto** [cilium/cilium#5317]
# Summary 

Found more than 70 messages of context-deadline in the logs. Endpoints are in not ready state. 

# Build information

Link: https://jenkins.cilium.io/job/cilium-ginkgo/job/cilium/job/master/1461/testReport/junit/k8s-1/11/K8sPolicyTest_Basic_Test_checks_all_kind_of_Kubernetes_policies/
Logs: 
[137e24ba_K8sPolicyTest_Basic_Test_checks_all_kind_of_Kubernetes_policies.zip](https://github.com/cilium/cilium/files/2314148/137e24ba_K8sPolicyTest_Basic_Test_checks_all_kind_of_Kubernetes_policies.zip)

Build Commit: 6fcef49c5b14186640466a67fa5ee30fb80a652e

# CI preview

```
Stacktrace

/home/jenkins/workspace/cilium-ginkgo_cilium_master-QVKDPVJ2A6E272IRUX7X6DCB5FBLCQDKRTOMHPZS7ZECX3RM5C3A/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
Cannot install "/home/vagrant/go/src/github.com/cilium/cilium/test/k8sT/manifests/l7-policy.yaml" policy
Expected
    <*helpers.SSHMetaError | 0xc420f97d60>: Error: Timeout reached: cannot change state of resource correctly; command timed out
Extended info: Policy version when cilium start
Cilium-agent "cilium-n8cvc" started with policy revision '10' and finished with '12' 
Cilium-agent "cilium-tg5js" started with policy revision '11' and finished with '13' 


to be nil
/home/jenkins/workspace/cilium-ginkgo_cilium_master-QVKDPVJ2A6E272IRUX7X6DCB5FBLCQDKRTOMHPZS7ZECX3RM5C3A/src/github.com/cilium/cilium/test/k8sT/Policies.go:217

Standard Output

⚠️  Number of "context deadline exceeded" in logs: 70
⚠️  Number of "level=error" in logs: 15
⚠️  Number of "level=warning" in logs: 34
Number of "Cilium API handler panicked" in logs: 0
Cilium pods: [cilium-n8cvc cilium-tg5js]
Netpols loaded: 
CiliumNetworkPolicies loaded: default::l7-policy 
Endpoint Policy Enforcement:
     app1-7c5785f7f7-jwg4n              =>   ingress
     app2-69f6b59665-c2hrb              =>   none
     app3-bd87968cd-zrkf4               =>   none
     coredns-687db6485c-89ktk           =>   none
     prometheus-core-6546477dc8-9l7cd   =>   none
     app1-7c5785f7f7-78vm9              =>   ingress
⚠️  Cilium agent "cilium-n8cvc": Status: Ok  Health: Ok Nodes "k8s1 k8s2" ContinerRuntime: Ok Kubernetes: Ok KVstore: Ok Controllers: Total 50 Failed 1
Failed controllers:
 controller sync-cnp-policy-status (v2 default/l7-policy) failure "context deadline exceeded"
Cilium agent "cilium-tg5js": Status: Ok  Health: Ok Nodes "k8s1 k8s2" ContinerRuntime: Ok Kubernetes: Ok KVstore: Ok Controllers: Total 16 Failed 0
Failed controllers:
 controller sync-cnp-policy-status (v2 default/l7-policy) failure "context deadline exceeded"

Standard Error

STEP: Testing L3/L4 rules
STEP: Testing L7 Policy
=== Test Finished at 2018-08-22T18:21:25Z====
===================== TEST FAILED =====================
cmd: kubectl get pods -o wide --all-namespaces
NAMESPACE     NAME                               READY     STATUS    RESTARTS   AGE       IP              NODE
default       app1-7c5785f7f7-78vm9              1/1       Running   0          17m       10.10.0.91      k8s1
default       app1-7c5785f7f7-jwg4n              1/1       Running   0          17m       10.10.0.20      k8s1
default       app2-69f6b59665-c2hrb              1/1       Running   0          17m       10.10.0.238     k8s1
default       app3-bd87968cd-zrkf4               1/1       Running   0          17m       10.10.0.16      k8s1
kube-system   cilium-n8cvc                       1/1       Running   0          19m       192.168.36.11   k8s1
kube-system   cilium-tg5js                       1/1       Running   0          19m       192.168.36.12   k8s2
kube-system   coredns-687db6485c-89ktk           1/1       Running   0          1h        10.10.0.234     k8s1
kube-system   etcd-k8s1                          1/1       Running   0          1h        192.168.36.11   k8s1
kube-system   kube-apiserver-k8s1                1/1       Running   0          1h        192.168.36.11   k8s1
kube-system   kube-controller-manager-k8s1       1/1       Running   0          1h        192.168.36.11   k8s1
kube-system   kube-proxy-ds4rc                   1/1       Running   0          1h        192.168.36.11   k8s1
kube-system   kube-proxy-kh7wh                   1/1       Running   0          1h        192.168.36.12   k8s2
kube-system   kube-scheduler-k8s1                1/1       Running   0          1h        192.168.36.11   k8s1
monitoring    prometheus-core-6546477dc8-9l7cd   1/1       Running   0          59m       10.10.0.7       k8s1

cmd: kubectl exec -n kube-system cilium-n8cvc -- cilium service list
ID   Frontend             Backend                   
1    10.96.0.1:443        1 => 192.168.36.11:6443   
2    10.96.0.10:53        1 => 10.10.0.234:53       
3    10.108.20.185:9090   1 => 10.10.0.7:9090       
4    10.109.122.50:80     1 => 10.10.0.20:80        
                          2 => 10.10.0.91:80        

cmd: kubectl exec -n kube-system cilium-n8cvc -- cilium endpoint list
ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                              IPv6                 IPv4          STATUS   
           ENFORCEMENT        ENFORCEMENT                                                                                                              
11999      Disabled           Disabled          16906      k8s:io.cilium.k8s.policy.serviceaccount=coredns          f00d::a0a:0:0:2edf   10.10.0.234   not-ready   
                                                           k8s:io.kubernetes.pod.namespace=kube-system                                                             
                                                           k8s:k8s-app=kube-dns                                                                                    
13924      Disabled           Disabled          7078       k8s:app=prometheus                                       f00d::a0a:0:0:3664   10.10.0.7     not-ready   
                                                           k8s:component=core                                                                                      
                                                           k8s:io.cilium.k8s.policy.serviceaccount=prometheus-k8s                                                  
                                                           k8s:io.kubernetes.pod.namespace=monitoring                                                              
32680      Enabled            Disabled          2501       k8s:id=app1                                              f00d::a0a:0:0:7fa8   10.10.0.20    not-ready   
                                                           k8s:io.cilium.k8s.policy.serviceaccount=app1-account                                                    
                                                           k8s:io.kubernetes.pod.namespace=default                                                                 
                                                           k8s:zgroup=testapp                                                                                      
34779      Disabled           Disabled          27774      k8s:id=app3                                              f00d::a0a:0:0:87db   10.10.0.16    not-ready   
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                                         
                                                           k8s:io.kubernetes.pod.namespace=default                                                                 
                                                           k8s:zgroup=testapp                                                                                      
46926      Disabled           Disabled          22907      k8s:appSecond=true                                       f00d::a0a:0:0:b74e   10.10.0.238   not-ready   
                                                           k8s:id=app2                                                                                             
                                                           k8s:io.cilium.k8s.policy.serviceaccount=app2-account                                                    
                                                           k8s:io.kubernetes.pod.namespace=default                                                                 
                                                           k8s:zgroup=testapp                                                                                      
53649      Disabled           Disabled          4          reserved:health                                          f00d::a0a:0:0:d191   10.10.0.8     not-ready   
58517      Enabled            Disabled          2501       k8s:id=app1                                              f00d::a0a:0:0:e495   10.10.0.91    not-ready   
                                                           k8s:io.cilium.k8s.policy.serviceaccount=app1-account                                                    
                                                           k8s:io.kubernetes.pod.namespace=default                                                                 
                                                           k8s:zgroup=testapp                                                                                      

cmd: kubectl exec -n kube-system cilium-tg5js -- cilium service list
ID   Frontend             Backend                   
1    10.96.0.1:443        1 => 192.168.36.11:6443   
2    10.96.0.10:53        1 => 10.10.0.234:53       
3    10.108.20.185:9090   1 => 10.10.0.7:9090       
4    10.109.122.50:80     1 => 10.10.0.20:80        
                          2 => 10.10.0.91:80        

cmd: kubectl exec -n kube-system cilium-tg5js -- cilium endpoint list
ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])   IPv6                   IPv4          STATUS   
           ENFORCEMENT        ENFORCEMENT                                                                                     
45789      Disabled           Disabled          4          reserved:health               f00d::a0a:100:0:b2dd   10.10.1.244   ready   
```

**Comment by tgraf** [comment]
It's because of proxy configuration timeouts:

```
2018-08-22T18:20:36.017192151Z level=error msg="destroying BPF maps due to errors during regeneration" containerID=74f4fe4909 endpointID=34779 error="proxy state changes failed: context deadline exce
eded" ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11
```

```
2018-08-22T18:20:36.01726451Z level=info msg="Regeneration of BPF program has completed" buildDuration=55.090674532s containerID=74f4fe4909 endpointID=34779 ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11
```

**Comment by rlenglet** [comment]
More detailed timeline:
```
$ grep 'endpointID=34779' pod-kube-system-cilium-n8cvc-cilium-agent.log | grep 'Regenerating BPF program\|Preparing to compile BPF\|Waiting for proxy updates to complete\|proxy state changes failed'
...
2018-08-22T18:19:40.913597834Z level=info msg="Regenerating BPF program" containerID=74f4fe4909 endpointID=34779 ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11 startTime="2018-08-22 18:19:40.909176426 +0000 UTC m=+970.308312968"
2018-08-22T18:19:40.959361864Z level=debug msg="Preparing to compile BPF" bpfHeaderfilesChanged=false containerID=74f4fe4909 endpointID=34779 ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11
2018-08-22T18:19:40.96118145Z level=debug msg="Waiting for proxy updates to complete..." containerID=74f4fe4909 endpointID=34779 ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11
2018-08-22T18:20:36.017192151Z level=error msg="destroying BPF maps due to errors during regeneration" containerID=74f4fe4909 endpointID=34779 error="proxy state changes failed: context deadline exceeded" ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11
2018-08-22T18:20:36.017320693Z level=warning msg="Regeneration of endpoint program failed" containerID=74f4fe4909 endpointID=34779 error="Error while configuring proxy redirects: proxy state changes failed: context deadline exceeded" ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11
2018-08-22T18:20:36.017336328Z level=debug msg="Error regenerating endpoint: Error while configuring proxy redirects: proxy state changes failed: context deadline exceeded" code=Failure containerID=74f4fe4909 endpointID=34779 endpointState=ready ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11 type=200
```

Envoy seems healthy. No error messages:
```
$ grep subsys=envoy pod-kube-system-cilium-n8cvc-cilium-agent.log  | grep -v debug | tail -n 5
2018-08-22T18:16:18.960903367Z level=info msg="[external/envoy/source/server/configuration_impl.cc:94] loading tracing configuration" subsys=envoy-config threadID=3654
2018-08-22T18:16:18.96092186Z level=info msg="[external/envoy/source/server/configuration_impl.cc:116] loading stats sink configuration" subsys=envoy-config threadID=3654
2018-08-22T18:16:18.962900016Z level=info msg="[external/envoy/source/server/server.cc:390] all clusters initialized. initializing init manager" subsys=envoy-main threadID=3654
2018-08-22T18:16:19.047277869Z level=info msg="[external/envoy/source/server/server.cc:410] starting main dispatch loop" subsys=envoy-main threadID=3654
2018-08-22T18:16:19.085260065Z level=info msg="[external/envoy/source/server/listener_manager_impl.cc:903] all dependencies initialized. starting workers" subsys=envoy-config threadID=3654
```

We see the NPDS cache get correctly updated with the new network policy for the pod (IP addresses `10.10.0.16` and `f00d::a0a:0:0:87db`:
```
$ grep subsys=xds pod-kube-system-cilium-n8cvc-cilium-agent.log
...
2018-08-22T18:19:40.954684426Z level=debug msg="preparing new cache transaction: upserting 1 entries, deleting 0 entries" subsys=xds xdsTypeURL=type.googleapis.com/cilium.NetworkPolicy xdsVersionInfo=111
2018-08-22T18:19:40.954777864Z level=debug msg="updating resource in cache" subsys=xds xdsResourceName="f00d::a0a:0:0:87db" xdsTypeURL=type.googleapis.com/cilium.NetworkPolicy xdsVersionInfo=111
2018-08-22T18:19:40.954963037Z level=debug msg="committing cache transaction and notifying of new version" subsys=xds xdsTypeURL=type.googleapis.com/cilium.NetworkPolicy xdsVersionInfo=111
2018-08-22T18:19:40.955362554Z level=debug msg="preparing new cache transaction: upserting 1 entries, deleting 0 entries" subsys=xds xdsTypeURL=type.googleapis.com/cilium.NetworkPolicy xdsVersionInfo=112
2018-08-22T18:19:40.956156206Z level=debug msg="updating resource in cache" subsys=xds xdsResourceName=10.10.0.16 xdsTypeURL=type.googleapis.com/cilium.NetworkPolicy xdsVersionInfo=112
2018-08-22T18:19:40.956190288Z level=debug msg="committing cache transaction and notifying of new version" subsys=xds xdsTypeURL=type.googleapis.com/cilium.NetworkPolicy xdsVersionInfo=112
```

However, we don't see Envoy query for it. We see only one xDS response, much earlier at 18:16:19.055, and that is only for listeners (not NPDS nor NPHDS):
```
$ grep subsys=xds pod-kube-system-cilium-n8cvc-cilium-agent.log  | grep 'sending xDS response'
2018-08-22T18:16:19.055832226Z level=debug msg="sending xDS response with 0 resources" subsys=xds xdsCanary=false xdsNonce=1 xdsStreamID=1 xdsTypeURL=type.googleapis.com/envoy.api.v2.Listener xdsVersionInfo=0
```

Envoy is actually misconfigured by the Cilium agent: it is configured with no listeners. That is why there are no Cilium filters configured, so nothing queries for NPDS nor NPHDS, which explains why regeneration times out waiting for an ACK:
```
$ grep subsys=xds pod-kube-system-cilium-n8cvc-cilium-agent.log  | grep 'xDS stream\|sending xDS response'
2018-08-22T18:16:19.051708908Z level=info msg="starting xDS stream processing" subsys=xds xdsStreamID=1
2018-08-22T18:16:19.05210549Z level=debug msg="received request from xDS stream" subsys=xds xdsClientNode="id:\"host~127.0.0.1~no-id~localdomain\" cluster:\"cluster1\" build_version:\"6fcef49c5b14186640466a67fa5ee30fb80a652e/1.8.0-dev/Distribution/RELEASE\" " xdsNonce= xdsStreamID=1 xdsTypeURL=type.googleapis.com/envoy.api.v2.Listener xdsVersionInfo=
2018-08-22T18:16:19.055832226Z level=debug msg="sending xDS response with 0 resources" subsys=xds xdsCanary=false xdsNonce=1 xdsStreamID=1 xdsTypeURL=type.googleapis.com/envoy.api.v2.Listener xdsVersionInfo=0
2018-08-22T18:16:19.087315121Z level=debug msg="received request from xDS stream" subsys=xds xdsClientNode="id:\"host~127.0.0.1~no-id~localdomain\" cluster:\"cluster1\" build_version:\"6fcef49c5b14186640466a67fa5ee30fb80a652e/1.8.0-dev/Distribution/RELEASE\" " xdsNonce=1 xdsStreamID=1 xdsTypeURL=type.googleapis.com/envoy.api.v2.Listener xdsVersionInfo=0
```

That's not right, since 2 listeners are supposed to be created:
```
$ grep subsys=envoy-manager pod-kube-system-cilium-n8cvc-cilium-agent.log  | grep 'addListener\|removeListener'
2018-08-22T18:16:18.13397554Z level=debug msg="Envoy: addListener 58517:ingress:TCP:80:11837" subsys=envoy-manager
2018-08-22T18:16:24.41699864Z level=debug msg="Envoy: addListener 32680:ingress:TCP:80:11298" subsys=envoy-manager
```

Those 2 listeners are:

- ingress to endpoint 58517, port TCP/80
- ingress to endpoint 32680, port TCP/80

Note that no endpoints are created for endpoint 34779. The `proxy state changes failed: context deadline exceeded` error message was not due to waiting for an ACK from Envoy. It was only indicating that BPF compilation took >55 seconds.

However, the fact that Envoy didn't receive the 2 configured listeners is an unrelated (but real) bug.

We see 489 updates to the xDS cache:
```
$ grep subsys=xds pod-kube-system-cilium-n8cvc-cilium-agent.log | grep 'preparing new cache transaction' | wc -l
     489
```
But all those updates are for NPDS and NPHDS, none are for listeners:
```
$ grep subsys=xds pod-kube-system-cilium-n8cvc-cilium-agent.log | grep 'preparing new cache transaction' | sed -e 's/^.*xdsTypeURL=\([^ ]*\).*$/\1/' | sort | uniq
type.googleapis.com/cilium.NetworkPolicy
type.googleapis.com/cilium.NetworkPolicyHosts
```
That is why we return no Listeners to Envoy. So both Envoy and the xDS layer are fine.

The issue is that the attempts to add the listeners failed due to contexts timing out:
```
$ grep 'context cancelled.*subsys=xds\|Envoy: addListener' pod-kube-system-cilium-n8cvc-cilium-agent.log
2018-08-22T18:16:18.13397554Z level=debug msg="Envoy: addListener 58517:ingress:TCP:80:11837" subsys=envoy-manager
2018-08-22T18:16:18.154817446Z level=warning msg="context cancelled: context deadline exceeded" subsys=xds xdsTypeURL=type.googleapis.com/envoy.api.v2.Listener
2018-08-22T18:16:24.41699864Z level=debug msg="Envoy: addListener 32680:ingress:TCP:80:11298" subsys=envoy-manager
2018-08-22T18:16:24.419010647Z level=warning msg="context cancelled: context deadline exceeded" subsys=xds xdsTypeURL=type.googleapis.com/envoy.api.v2.Listener
```

So there are 2 issues to address here:

1. In the 3 cases of timeouts, BPF regeneration took more 55 seconds to get to the end of the BPF compilation and re-acquire the endpoint mutex.
2. We don't properly handle context timeouts in the code that creates listeners during BPF regeneration. We should delete any newly created listener if we find out that the context has timed out. Or we should ignore context timeouts when upserting or deleting from the xDS cache. I will do the latter, as it's much simpler and has no negative consequences.

I'll send a PR to fix 2.

**Comment by tgraf** [comment]
Great analysis

**Comment by tgraf** [comment]
> Note that no endpoints are created for endpoint 34779. The proxy state changes failed: context deadline exceeded error message was not due to waiting for an ACK from Envoy. It was only indicating that BPF compilation took >55 seconds.

This cannot be accurate as no compilation was performed as we would see a log message as per:
```
                err = e.runInit(libdir, rundir, epdir, epInfoCache.ifName, debug)
                e.getLogger().WithError(err).
                        WithField(logfields.BPFCompilationTime, time.Since(start).String()).
                        Debugf("BPF compilation completed")
```

The timestamps also make it obvious that we are waiting within `WaitForProxyCompletions`:
```
2018-08-22T18:19:40.96118145Z level=debug msg="Waiting for proxy updates to complete..." containerID=74f4fe4909 endpointID=34779 ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11
2018-08-22T18:20:36.017192151Z level=error msg="destroying BPF maps due to errors during regeneration" containerID=74f4fe4909 endpointID=34779 error="proxy state changes failed: context deadline exceeded" ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11
```

**Comment by rlenglet** [comment]
@tgraf I was referring to an earlier regeneration where the context timed out before we add/remove listeners and before we start waiting for Envoy's ACK:
```
2018-08-22T18:14:11.324084718Z level=info msg="Regenerating BPF program" containerID=8a5bf4540a endpointID=58517 ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=10 startTime="2018-08-22 18:14:11.322691866 +0000 UTC m=+640.721828389"
...
2018-08-22T18:16:17.658451314Z level=debug msg="BPF compilation completed" BPFCompilationTime=2m6.029706074s containerID=8a5bf4540a endpointID=58517 error="<nil>" ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=10
...
2018-08-22T18:16:18.13397554Z level=debug msg="Envoy: addListener 58517:ingress:TCP:80:11837" subsys=envoy-manager
2018-08-22T18:16:18.154817446Z level=warning msg="context cancelled: context deadline exceeded" subsys=xds xdsTypeURL=type.googleapis.com/envoy.api.v2.Listener
...
2018-08-22T18:16:18.15730783Z level=debug msg="Waiting for proxy updates to complete..." containerID=8a5bf4540a endpointID=58517 ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=10
2018-08-22T18:16:18.158296428Z level=error msg="destroying BPF maps due to errors during regeneration" containerID=8a5bf4540a endpointID=58517 error="proxy state changes failed: context deadline exceeded" ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=10
```
The problem when that happens is that the endpoint's state re: realized listeners is inconsistent with the xDS cache. That's the issue that is fixed in https://github.com/cilium/cilium/pull/5326.

**Comment by tgraf** [comment]
@rlenglet That seems to be for a different endpoint ID though, 58517 now vs 34779 before. I understand why 58517 broke and how your change applies there. What happened to 34779?

**Comment by tgraf** [comment]
Reopening as not all issues reported by this issue have been resolved yet.

**Comment by rlenglet** [comment]
In cilium-n8cvc, all endpoints are `not-ready`, all because of context deadline exceeded during regeneration:
```
$ jq '.[] | {id: .id, state: .status.state, log: .status.log}' cilium-n8cvc_endpoint_list.txt 
{
  "id": 11999,
  "state": "not-ready",
  "log": [
    {
      "code": "Failure",
      "message": "Error regenerating endpoint: Error while configuring proxy redirects: proxy state changes failed: context deadline exceeded",
      "state": "ready",
      "timestamp": "2018-08-22T18:20:36Z"
    }
  ]
}
{
  "id": 13924,
  "state": "not-ready",
  "log": [
    {
      "code": "Failure",
      "message": "Error regenerating endpoint: Error while configuring proxy redirects: proxy state changes failed: context deadline exceeded",
      "state": "ready",
      "timestamp": "2018-08-22T18:21:31Z"
    }
  ]
}
{
  "id": 32680,
  "state": "not-ready",
  "log": [
    {
      "code": "Failure",
      "message": "Error regenerating endpoint: Error while configuring proxy redirects: proxy state changes failed: context deadline exceeded",
      "state": "ready",
      "timestamp": "2018-08-22T18:19:40Z"
    }
  ]
}
{
  "id": 34779,
  "state": "not-ready",
  "log": [
    {
      "code": "Failure",
      "message": "Error regenerating endpoint: Error while configuring proxy redirects: proxy state changes failed: context deadline exceeded",
      "state": "ready",
      "timestamp": "2018-08-22T18:20:36Z"
    }
  ]
}
{
  "id": 46926,
  "state": "not-ready",
  "log": [
    {
      "code": "Failure",
      "message": "Error regenerating endpoint: Error while configuring proxy redirects: proxy state changes failed: context deadline exceeded",
      "state": "ready",
      "timestamp": "2018-08-22T18:21:31Z"
    }
  ]
}
{
  "id": 53649,
  "state": "not-ready",
  "log": [
    {
      "code": "Failure",
      "message": "Error regenerating endpoint: Error while configuring proxy redirects: proxy state changes failed: context deadline exceeded",
      "state": "ready",
      "timestamp": "2018-08-22T18:22:26Z"
    }
  ]
}
{
  "id": 58517,
  "state": "not-ready",
  "log": [
    {
      "code": "Failure",
      "message": "Error regenerating endpoint: Error while configuring proxy redirects: proxy state changes failed: context deadline exceeded",
      "state": "ready",
      "timestamp": "2018-08-22T18:19:40Z"
    }
  ]
}
```

Note that only endpoints 32680 and 58517 (the ones for which we saw Envoy redirects getting created) have a policy enforced:
```
$ jq '.[] | {id: .id, ingress: .spec.options["IngressPolicy"], egress: .spec.options["EgressPolicy"]}' cilium-n8cvc_endpoint_list.txt 
{
  "id": 11999,
  "ingress": "Disabled",
  "egress": "Disabled"
}
{
  "id": 13924,
  "ingress": "Disabled",
  "egress": "Disabled"
}
{
  "id": 32680,
  "ingress": "Enabled",
  "egress": "Disabled"
}
{
  "id": 34779,
  "ingress": "Disabled",
  "egress": "Disabled"
}
{
  "id": 46926,
  "ingress": "Disabled",
  "egress": "Disabled"
}
{
  "id": 53649,
  "ingress": "Disabled",
  "egress": "Disabled"
}
{
  "id": 58517,
  "ingress": "Enabled",
  "egress": "Disabled"
}
```

There were 62 BPF regenerations, of which 15 took more than 55 seconds:
```
$ grep 'Regeneration of BPF program has completed' pod-kube-system-cilium-n8cvc-cilium-agent.log | sed -e 's/^.*buildDuration=\([^ ]*\).*$/\1/' | wc -l
      62
```
```
$ grep 'Regeneration of BPF program has completed' pod-kube-system-cilium-n8cvc-cilium-agent.log | sed -e 's/^[^T]*T\([^.]*\).*buildDuration=\([^ ]*\).*endpointID=\([^ ]*\).*$/\1 \2 \3/' | grep -v 'ms '
18:04:02 16.784966526s 13924
18:04:02 16.752224112s 11999
18:04:12 9.459511391s 53649
18:04:46 16.47983778s 46926
18:04:47 17.569262548s 34779
18:05:03 15.718919775s 32680
18:05:07 19.880536756s 58517
18:05:46 39.497144438s 46926
18:05:48 40.88923881s 34779
18:09:16 2m15.325336905s 58517
18:09:25 2m24.525038269s 32680
18:11:55 1m19.846516041s 58517
18:11:56 1m18.770317851s 32680
18:13:47 1.365283745s 11999
18:13:47 1.516452705s 58517
18:16:18 2m6.835577874s 58517
18:16:24 2m12.50815674s 32680
18:19:40 55.006956782s 58517
18:19:40 55.012410435s 32680
18:20:36 55.090674532s 34779
18:20:36 55.058222333s 11999
18:21:31 55.147854236s 13924
18:21:31 55.190592394s 46926
18:22:26 55.007241143s 53649
18:24:30 55.257234144s 46926
18:24:30 55.044578403s 58517
```

The first very suspicious regenerations (2m15 and 2m24) happen for endpoints 58517 and 32680, at 18:09:16 and 18:09:25. Note that the policies regenerated here are probably not L7, since we see no Envoy listeners getting created; the listeners for those endpoints get created in a later regeneration, at 18:16:18 and 18:16:24:
```
$ grep addListener pod-kube-system-cilium-n8cvc-cilium-agent.log 
2018-08-22T18:16:18.13397554Z level=debug msg="Envoy: addListener 58517:ingress:TCP:80:11837" subsys=envoy-manager
2018-08-22T18:16:24.41699864Z level=debug msg="Envoy: addListener 32680:ingress:TCP:80:11298" subsys=envoy-manager
```

**Comment by rlenglet** [comment]
I'll first focus on endpoint 58517, since it is the first endpoint to experience a regeneration timeout.
Its first 2 regenerations are successful:
```
$ grep 'Regeneration of BPF program has completed' pod-kube-system-cilium-n8cvc-cilium-agent.log | sed -e 's/^[^T]*T\([^.]*\).*buildDuration=\([^ ]*\).*endpointID=\([^ ]*\).*$/\1 \2 \3/' | grep ' 58517$'
18:05:07 19.880536756s 58517
18:05:59 114.408829ms 58517
18:09:16 2m15.325336905s 58517
18:09:26 206.285092ms 58517
18:11:55 1m19.846516041s 58517
18:13:47 1.516452705s 58517
18:16:18 2m6.835577874s 58517
18:19:40 55.006956782s 58517
18:24:30 55.044578403s 58517
```

Then the 3rd one times out:
```
$ grep 'Regenerating BPF program\|destroying BPF maps due to errors during regeneration' pod-kube-system-cilium-n8cvc-cilium-agent.log  | grep endpointID=58517 | more
...
2018-08-22T18:07:00.922686682Z level=info msg="Regenerating BPF program" containerID=8a5bf4540a endpointID=58517 ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=6 startTime="2018-08-22 18:07:00.92072208 +0000 UTC m=+210.319858598"
2018-08-22T18:09:16.246075551Z level=error msg="destroying BPF maps due to errors during regeneration" containerID=8a5bf4540a endpointID=58517 error="proxy state changes failed: context deadline exceeded" ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=6
```

The `BPF compilation completed` log message indicates an actual BPF compilation time of 2m15, and all subsequent compilations exceed the timeout:
```
$ grep 'BPF compilation completed.*endpointID=58517' pod-kube-system-cilium-n8cvc-cilium-agent.log 
...
2018-08-22T18:09:16.244376896Z level=debug msg="BPF compilation completed" BPFCompilationTime=2m15.126676385s containerID=8a5bf4540a endpointID=58517 error="<nil>" ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=6
2018-08-22T18:11:55.831092773Z level=debug msg="BPF compilation completed" BPFCompilationTime=1m18.895646343s containerID=8a5bf4540a endpointID=58517 error="<nil>" ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=8
2018-08-22T18:16:17.658451314Z level=debug msg="BPF compilation completed" BPFCompilationTime=2m6.029706074s containerID=8a5bf4540a endpointID=58517 error="<nil>" ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=10
```

So all the time was spent in BPF compilation in the regeneration.

Note that the `UpdateLabels` method for this endpoint is called concurrently with the BPF compilation, cf. the `Refreshing labels of endpoint` log:
```
2018-08-22T18:07:01.100591904Z level=debug msg="BPF header file hashed (was: \"2417dee0205e87557d95aa6cd8467ca1\")" bpfHeaderfileHash=c517b8bcbfb58002a828f13fae5a9318 containerID=8a5bf4540a endpointID=58517 ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=6
2018-08-22T18:07:01.115211358Z level=debug msg="Preparing to compile BPF" bpfHeaderfilesChanged=true containerID=8a5bf4540a endpointID=58517 ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=6
2018-08-22T18:07:18.533146456Z level=debug msg="Refreshing labels of endpoint" containerID=8a5bf4540a endpointID=58517 identityLabels="k8s:id=app1,k8s:io.cilium.k8s.policy.serviceaccount=app1-account,k8s:io.kubernetes.pod.namespace=default,k8s:zgroup=testapp" infoLabels="container:annotation.kubernetes.io/config.seen=2018-08-22T18:04:26.215739934Z,container:annotation.kubernetes.io/config.source=api,container:io.kubernetes.container.name=POD,container:io.kubernetes.docker.type=podsandbox,container:io.kubernetes.pod.name=app1-7c5785f7f7-78vm9,container:io.kubernetes.pod.uid=ce57342d-a635-11e8-b11a-0800275579e8,k8s:pod-template-hash=3713419393" subsys=endpoint
2018-08-22T18:07:32.88221881Z level=debug msg="Successfully annotated pod with cilium.io/identity=2501" endpointID=58517 k8sIdentityAnnotation=cilium.io/identity k8sNamespace=default k8sPodName=app1-7c5785f7f7-78vm9 retryUUID=3d8ea212-a636-11e8-93ff-0800275579e8 subsys=k8s
2018-08-22T18:07:47.952410957Z level=debug msg="Refreshing labels of endpoint" containerID=8a5bf4540a endpointID=58517 identityLabels="k8s:id=app1,k8s:io.cilium.k8s.policy.serviceaccount=app1-account,k8s:io.kubernetes.pod.namespace=default,k8s:zgroup=testapp" infoLabels="container:annotation.kubernetes.io/config.seen=2018-08-22T18:04:26.215739934Z,container:annotation.kubernetes.io/config.source=api,container:io.kubernetes.container.name=POD,container:io.kubernetes.docker.type=podsandbox,container:io.kubernetes.pod.name=app1-7c5785f7f7-78vm9,container:io.kubernetes.pod.uid=ce57342d-a635-11e8-b11a-0800275579e8,k8s:pod-template-hash=3713419393" subsys=endpoint
2018-08-22T18:09:16.244376896Z level=debug msg="BPF compilation completed" BPFCompilationTime=2m15.126676385s containerID=8a5bf4540a endpointID=58517 error="<nil>" ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=6
2018-08-22T18:09:16.245552735Z level=debug msg="Waiting for proxy updates to complete..." containerID=8a5bf4540a endpointID=58517 ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=6
```
However, this may only be a red herring. This method is locking the endpoint a few times, but I don't think that it's locking for too long. The operations that may take time are performed in the resolve-identity-* controllers, which do lock the endpoint, but don't seem to take long to run in this case:
```
$ grep 'Controller func execution time.*resolve-identity-58517' pod-kube-system-cilium-n8cvc-cilium-agent.log
2018-08-22T18:04:29.901172951Z level=debug msg="Controller func execution time: 15.90133ms" name=resolve-identity-58517 subsys=controller uuid=d092864d-a635-11e8-93ff-0800275579e8
2018-08-22T18:04:32.075276658Z level=debug msg="Controller func execution time: 58.721241ms" name=resolve-identity-58517 subsys=controller uuid=d092864d-a635-11e8-93ff-0800275579e8
2018-08-22T18:09:32.085374962Z level=debug msg="Controller func execution time: 653.763µs" name=resolve-identity-58517 subsys=controller uuid=d092864d-a635-11e8-93ff-0800275579e8
2018-08-22T18:14:32.1307245Z level=debug msg="Controller func execution time: 11.193373ms" name=resolve-identity-58517 subsys=controller uuid=d092864d-a635-11e8-93ff-0800275579e8
2018-08-22T18:19:32.17070662Z level=debug msg="Controller func execution time: 13.383782ms" name=resolve-identity-58517 subsys=controller uuid=d092864d-a635-11e8-93ff-0800275579e8
2018-08-22T18:24:32.184466256Z level=debug msg="Controller func execution time: 460.304µs" name=resolve-identity-58517 subsys=controller uuid=d092864d-a635-11e8-93ff-0800275579e8
```
```
$ grep 'Controller func execution time.*resolve-identity-32680' pod-kube-system-cilium-n8cvc-cilium-agent.log
2018-08-22T18:04:29.963467995Z level=debug msg="Controller func execution time: 16.74512ms" name=resolve-identity-32680 subsys=controller uuid=d0ad77f2-a635-11e8-93ff-0800275579e8
2018-08-22T18:04:31.854157315Z level=debug msg="Controller func execution time: 223.943494ms" name=resolve-identity-32680 subsys=controller uuid=d0ad77f2-a635-11e8-93ff-0800275579e8
2018-08-22T18:09:31.865390955Z level=debug msg="Controller func execution time: 2.460329ms" name=resolve-identity-32680 subsys=controller uuid=d0ad77f2-a635-11e8-93ff-0800275579e8
2018-08-22T18:14:31.888797411Z level=debug msg="Controller func execution time: 1.579728ms" name=resolve-identity-32680 subsys=controller uuid=d0ad77f2-a635-11e8-93ff-0800275579e8
2018-08-22T18:19:31.901392568Z level=debug msg="Controller func execution time: 393.549µs" name=resolve-identity-32680 subsys=controller uuid=d0ad77f2-a635-11e8-93ff-0800275579e8
2018-08-22T18:24:31.963823154Z level=debug msg="Controller func execution time: 21.975958ms" name=resolve-identity-32680 subsys=controller uuid=d0ad77f2-a635-11e8-93ff-0800275579e8
```

**Comment by rlenglet** [comment]
Note that the first two BPF compilations that timed out were concurrent:
```
2018-08-22T18:07:01.115211358Z level=debug msg="Preparing to compile BPF" bpfHeaderfilesChanged=true containerID=8a5bf4540a endpointID=58517 ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=6
2018-08-22T18:07:01.499588854Z level=debug msg="Preparing to compile BPF" bpfHeaderfilesChanged=true containerID=3327623e80 endpointID=32680 ipv4=10.10.0.20 ipv6="f00d::a0a:0:0:7fa8" k8sPodName=default/app1-7c5785f7f7-jwg4n policyRevision=6
2018-08-22T18:09:16.244376896Z level=debug msg="BPF compilation completed" BPFCompilationTime=2m15.126676385s containerID=8a5bf4540a endpointID=58517 error="<nil>" ipv4=10.10.0.91 ipv6="f00d::a0a:0:0:e495" k8sPodName=default/app1-7c5785f7f7-78vm9 policyRevision=6
2018-08-22T18:09:25.637759575Z level=debug msg="BPF compilation completed" BPFCompilationTime=2m24.140848543s containerID=3327623e80 endpointID=32680 error="<nil>" ipv4=10.10.0.20 ipv6="f00d::a0a:0:0:7fa8" k8sPodName=default/app1-7c5785f7f7-jwg4n policyRevision=6
```
So it appears that something independent from the endpoints or even the Cilium Agent started using CPU and caused these timeouts. I would assume something in the environment.

`cilium-node-monitor` starts reading the perf buffer during that time, at 18:07:24:
```
$ grep subsys=cilium-node-monitor pod-kube-system-cilium-n8cvc-cilium-agent.log 
2018-08-22T18:03:44.065391164Z level=info msg="Serving cilium node monitor v1.0 API at unix:///var/run/cilium/monitor.sock" subsys=cilium-node-monitor
2018-08-22T18:03:44.067025704Z level=info msg="Serving cilium node monitor v1.2 API at unix:///var/run/cilium/monitor1_2.sock" subsys=cilium-node-monitor
2018-08-22T18:03:44.067838762Z level=info msg="Beginning to read cilium agent events" subsys=cilium-node-monitor
2018-08-22T18:07:24.341521051Z level=info msg="New listener connected." count.listener=1 subsys=cilium-node-monitor version=1.2
2018-08-22T18:07:24.34226253Z level=info msg="Beginning to read perf buffer" startTime="2018-08-22 18:07:24.341768056 +0000 UTC m=+221.005567988" subsys=cilium-node-monitor
2018-08-22T18:21:40.497109736Z level=info msg="Listener disconnected" subsys=cilium-node-monitor
2018-08-22T18:21:40.497830135Z level=info msg="Removed listener" count.listener=0 subsys=cilium-node-monitor version=1.2
2018-08-22T18:21:40.527869122Z level=info msg="Stopped reading perf buffer" startTime="2018-08-22 18:07:24.341768056 +0000 UTC m=+221.005567988" subsys=cilium-node-monitor
```

When `cilium-node-monitor` starts reading the perf buffer, it may spend lots of CPU reading and decoding the events that have accumulated in the buffer.

**My conclusion is that those two first BPF compilation timeouts are due to the node monitor starting to read from the perf buffer concurrently.**

**Comment by rlenglet** [comment]
Now switching to endpoint 34779. Starting from 18:19:40, all endpoint regenerations start failing with an Envoy proxy configuration timeout. Even endpoints that have no policy enforced. Endpoint 34779 is the first endpoint with no policy enforced to see such a timeout.
```
$ grep 'Regeneration of BPF program has completed' pod-kube-system-cilium-n8cvc-cilium-agent.log | sed -e 's/^[^T]*T\([^.]*\).*buildDuration=\([^ ]*\).*endpointID=\([^ ]*\).*$/\1 \2 \3/' | tail -n 13
18:14:11 150.767794ms 46926
18:14:11 75.16705ms 13924
18:16:18 2m6.835577874s 58517
18:16:24 2m12.50815674s 32680
18:19:40 55.006956782s 58517
18:19:40 55.012410435s 32680
18:20:36 55.090674532s 34779
18:20:36 55.058222333s 11999
18:21:31 55.147854236s 13924
18:21:31 55.190592394s 46926
18:22:26 55.007241143s 53649
18:24:30 55.257234144s 46926
18:24:30 55.044578403s 58517
```

In the failing regeneration, the BPF program is not recompiled. We see 55 seconds elapse between the BPF file hash and the timeout. So all the time is spent waiting for an ACK from Envoy, which should not happen:
```
$ grep 'BPF compilation completed\|BPF header file unchanged, skipping BPF compilation and installation\|Regeneration of BPF program has completed' pod-kube-system-cilium-n8cvc-cilium-agent.log | grep endpointID=34779
...
2018-08-22T18:19:40.959609247Z level=debug msg="BPF header file unchanged, skipping BPF compilation and installation" bpfHeaderfileHash=c88afa476dde97f32008048b34c7b106 containerID=74f4fe4909 endpointID=34779 ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11
2018-08-22T18:20:36.01726451Z level=info msg="Regeneration of BPF program has completed" buildDuration=55.090674532s containerID=74f4fe4909 endpointID=34779 ipv4=10.10.0.16 ipv6="f00d::a0a:0:0:87db" k8sPodName=default/app3-bd87968cd-zrkf4 policyRevision=11
```

**Comment by rlenglet** [comment]
The cause for all the proxy timeouts was actually the inconsistency between the endpoint and proxy manager state. That has been fixed in https://github.com/cilium/cilium/pull/5326.

Here's what happened:

1. Due to compilation timeouts, at 18:16:18 and 18:16:24, the Cilium agent and the xDS cache become inconsistent. The agent believes that 2 listeners have been created, for endpoints 58517 and 32680. The 2 listeners were accounted for in both the Envoy manager state (`XDSServer.listeners`) and the endpoint's state (`Endpoint.realizedRedirects`), whereas they were actually not created in the xDS cache due to the timeout handling bug fixed by https://github.com/cilium/cilium/pull/5326.
2. Each subsequent regeneration results in a call to `Endpoint.updateNetworkPolicy`, which in turn calls `XDSServer.UpdateNetworkPolicy` which updates the xDS cache with the updated network policies. The logic in `XDSServer.UpdateNetworkPolicy` is that if any listener has been created, then Envoy will query for that network policy update, so we'll add a completion in order to wait an ACK for the network policy later in `regenerateBPF`. Cf. https://github.com/cilium/cilium/blob/6fcef49c5b14186640466a67fa5ee30fb80a652e/pkg/envoy/server.go#L590-L595 In this case, we believe (wrongly) that 2 listeners have been created, so we always add a completion for every call to `XDSServer.UpdateNetworkPolicy`.
3. In `regenerateBPF`, we'll wait on the completion / ACK, which never arrives because no listener has actually been created.

**So the root cause of those 55-second regeneration timeouts was the inconsistency between the endpoint and Envoy manager states, and the xDS cache. That bug has been fixed in https://github.com/cilium/cilium/pull/5326.**

**The earlier BPF compilation timeout issues are due to external resource constraints, e.g. CPU contention with `cilium-node-monitor`.**

**There is nothing more to fix. Closing this issue.**

**Comment by tgraf** [comment]
Is #5326 really a complete fix? We don't really know that the listener exists until Envoy has ACKed the change. Wouldn't it be a lot more reliable to always remove a newly created redirect from `e.realizedRedirect` if we are not receiving an ACK?

We have similar logic for BPF maps where newly created BPF maps are deleted again i a deferred function if the regeneration fails.

**Comment by rlenglet** [comment]
> Is #5326 really a complete fix? We don't really know that the listener exists until Envoy has ACKed the change. Wouldn't it be a lot more reliable to always remove a newly created redirect from e.realizedRedirect if we are not receiving an ACK?

In short, until we receive an ACK from Envoy for any revision, we can't know in what state Envoy is.
What happens if we don't receive an ACK for the 2nd update (to revert)? If the 1st update didn't succeed, and we didn't receive an ACK for it, there's a high chance that we won't receive an ACK for the 2nd update.
So I don't think this will work in practice.

**Comment by tgraf** [comment]
We can't just leave it unfixed. The first major problem is having state in `e.realizedRedirects` that is inaccurate as we never recover from it and the endpoint will be inaccessible forever. That can be avoided by only populating `e.realizedRedirects` if we have received an ACK.

**Comment by rlenglet** [comment]
`e.realizedRedirects` represents what is in the xDS cache. We need to keep that information. I'd be ok with renaming it, and adding a field `e.realizedRedirects` to track the ACKed redirects. I'm not sure what that would give us, though.

**Comment by rlenglet** [comment]
Now, I think there would be value in reverting xDS cache updates & waiting for an Envoy ACK in the case the BPF compilation times out, like we've seen in this issue.
In the case we time out while waiting for an ACK from Envoy, we can also revert the state in the xDS cache, but I wouldn't wait for an ACK in this case.

**Comment by rlenglet** [comment]
I could also try moving all xDS cache updates until after BPF compilation, and do that only if we're sure that BPF compilation has not timed out. That would reduce the need to revert xDS cache updates.

One drawback is that it may make regeneration longer, since we wouldn't parallelize the start of Envoy and its configuration / ACK.

**Comment by tgraf** [comment]
I'm not very opinionatd on the actual implementation but it has to be reliable and eventually consistent. Being in a situation where we wait for an ACK from an unconfigured Envoy is unacceptable.

**Comment by rlenglet** [comment]
My take would be:
- Refactor the wait for Envoy ACK in regeneration so that the completion wait group’s context doesn’t have a timeout anymore. The timeout would be purely local to the call to WaitForCompletion... That way, the time taken to recompile BPF won’t be counted towards the Envoy ACK wait anymore.
- In case of an Envoy ACK timeout, revert the changes to Listeners and NetworkPolicy in the xDS cache. But don’t bother waiting for an ACK for the revert, since Envoy is already not responsive.
- In case we receive a NACK for the xDS cache updates, revert the changes like above, but try to wait for an ACK in this case.

**Comment by rlenglet** [comment]
And in case of a NACK for a Listener creation due to a port conflict, we should quarantine the port, as Jarno proposed.
